libraries/CurieBLE/src/BLEProfile.cpp

```diff
@@ -0,0 +1,513 @@
+    struct bt_conn *conn = bt_conn_lookup_addr_le(_peripheral->bt_le_address());
+    
+    if (NULL == conn)
+    {
+        // Link lost
+        pr_debug(LOG_MODULE_APP, "Can't find connection\n");
+        return;
+    }
+    
+    // Reset start handle
+    _discover_params.start_handle = 0x0001;
+    serviceattr->discover(&_discover_params);
+    
+    err = bt_gatt_discover(conn, &_discover_params);
+    bt_conn_unref(conn);
+    if (err)
+    {
+        pr_debug(LOG_MODULE_APP, "Discover failed(err %d)\n", err);
+        return;
+    }
+}
+
+BLEAttribute *BLEProfile::attribute(struct bt_gatt_subscribe_params *params)
+{
+    return attribute(params->value_handle);
+}
+
+BLEAttribute *BLEProfile::attribute(const struct bt_uuid* uuid)
+{
+    int i;
+    BLEAttribute *attr_tmp = NULL;
+    BLECharacteristic *chrc_tmp = NULL;
+    bool att_found = false;
+    
+    for (i = 0; i < _num_attributes; i++)
+    {
+        attr_tmp = _attributes[i];
+        if ((NULL == attr_tmp) || (attr_tmp->type() != BLETypeCharacteristic))
+        {
+            continue;
+        }
+        chrc_tmp = (BLECharacteristic *)attr_tmp;
+        if (chrc_tmp->uuid() == uuid);
+        {
+            att_found = true;
+            break;
+        }
+    }
+    
+    if (false == att_found)
+    {
+        pr_debug(LOG_MODULE_APP, "Attributes not found");
+        // Didn't found the characteristic
+        chrc_tmp = NULL;
+    }
+    return chrc_tmp;
+}
+
+
+BLEAttribute *BLEProfile::attribute(uint16_t handle)
+{
+    int i;
+    struct bt_gatt_attr *attr_gatt = NULL;
+    for (i = 0; i < _attr_index; i++)
+    {
+        attr_gatt = _attr_base + i;
+        if (handle == attr_gatt->handle)
+        {
+            break;
+        }
+    }
+    
+    if (i < _attr_index && i > 1)
+    {
+        // Found the GATT ATTR
+        //  Serach the attribute 
+        //   Characteristic Declaration
+        //   Characteristic Descriptor
+        //   CCCD
+        attr_gatt--;
+        if (attr_gatt->uuid == BLECharacteristic::getCharacteristicAttributeUuid())
+        {
+            attr_gatt++;
+        }
+        else
+        {
+            attr_gatt--;
+            if (attr_gatt->uuid == BLECharacteristic::getCharacteristicAttributeUuid())
+            {
+                attr_gatt++;
+            }
+            else
+            {
+                attr_gatt = NULL;
+            }
+        }
+    }
+    else
+    {
+        attr_gatt = NULL;
+    }
+    
+    if (NULL != attr_gatt)
+    {
+        return attribute(attr_gatt->uuid);
+    }
+    return NULL;
+}
+
+
+void BLEProfile::clearHandles(void)
+{
+    int i;
+    struct bt_gatt_attr *attr = NULL;
+    // Didn't need to unsubscribe
+    //  The stack will unsubscribe the notify when disconnected.
+    //  The sub_param has some pointer. So can't call memset. Just reset the index.
+    //memset(_sub_param, 0x00, ((BLECharacteristic::numNotifyAttributes()) * sizeof(struct bt_gatt_subscribe_params)));
```

Comment:
Is this commented memset call part of a TODO for future work? If so, maybe just put an extra note specifying this. If not, can you remove it?